File: functions/url_key.py
"""
Canonical dedupe key for a saved URL (``urlKey`` on a link doc).

Dedupe used to be an exact string match on ``url``, so the same article saved
as ``http://www.site.com/a/``, ``https://site.com/a?utm_source=x`` and
``https://m.site.com/a#top`` became three cards. ``url_key`` folds those to one
key; the stored ``url`` stays exactly what the user shared.

MIRRORED IN ``web/lib/urlKey.ts`` (``urlKey``). The web form dedupes against the
same field before it writes its placeholder, so the two MUST produce identical
keys. tests/test_url_key.py pins the cases; change both files together.

Rules (in order):
- scheme: http and https are the same page → always ``https``
- host: lower-cased, default port dropped, a leading ``www.`` / ``m.`` /
  ``mobile.`` stripped, ``twitter.com`` → ``x.com``
- YouTube: ``youtu.be/ID``, ``/shorts/ID``, ``/live/ID``, ``/embed/ID`` and
  ``watch?v=ID&…`` all become ``youtube.com/watch?v=ID`` (every other param
  — t, si, feature, list, pp — dropped)
- fragment dropped
- trailing ``/`` dropped (the bare root becomes no path at all)
- tracking params dropped (utm_*, fbclid, gclid, mc_*, igsh, ref, …; ``si`` on
  YouTube/Spotify; ``s``/``t`` on x.com); the rest are kept, sorted, so param
  order doesn't split a page in two
"""

from typing import Optional
from urllib.parse import urlsplit, parse_qsl, urlencode, quote

# Dropped on every host. Exact names; anything starting with a prefix in
# _TRACKING_PREFIXES is dropped too.
_TRACKING_PARAMS = frozenset({
    "fbclid", "gclid", "dclid", "gbraid", "wbraid", "msclkid", "yclid",
    "twclid", "ttclid", "li_fat_id", "igsh", "igshid", "ref", "ref_src",
    "ref_url", "_ga", "_gl", "mkt_tok", "oly_anon_id", "oly_enc_id",
    "vero_id", "wickedid", "rb_clickid", "s_cid", "__s", "_hsenc", "_hsmi",
    "spm", "share_id", "sharesource", "cmpid",
})
_TRACKING_PREFIXES = ("utm_", "mc_", "pk_", "hsa_")

# Per-host extras: short names that are tracking on these hosts but may be real
# parameters elsewhere (a search site's `s=` is the query).
_HOST_TRACKING = {
    "x.com": frozenset({"s", "t"}),
    "youtube.com": frozenset({"si", "feature", "pp"}),
    "open.spotify.com": frozenset({"si", "context", "nd"}),
    "spotify.com": frozenset({"si"}),
    "instagram.com": frozenset({"img_index"}),
}

_HOST_PREFIXES = ("www.", "m.", "mobile.")
_HOST_ALIASES = {"twitter.com": "x.com"}

_YOUTUBE_HOSTS = ("youtube.com", "youtube-nocookie.com")
_YOUTUBE_PATH_IDS = ("/shorts/", "/live/", "/embed/", "/v/")


def _clean_host(host: str) -> str:
    host = (host or "").strip().lower().rstrip(".")
    for prefix in _HOST_PREFIXES:
        if host.startswith(prefix) and host.count(".") >= 2:
            host = host[len(prefix):]
            break
    return _HOST_ALIASES.get(host, host)


def _valid_yt_id(value: str) -> Optional[str]:
    value = (value or "").strip()
    if 6 <= len(value) <= 20 and all(c.isalnum() or c in "-_" for c in value):
        return value
    return None


def _youtube_id(host: str, path: str, params: list) -> Optional[str]:
    if host == "youtu.be":
        return _valid_yt_id(path.strip("/").split("/", 1)[0])
    if host in _YOUTUBE_HOSTS:
        if path.rstrip("/") == "/watch":
            for k, v in params:
                if k == "v":
                    return _valid_yt_id(v)
        for prefix in _YOUTUBE_PATH_IDS:
            if path.startswith(prefix):
                return _valid_yt_id(path[len(prefix):].split("/", 1)[0])
    return None


def _is_tracking(key: str, host: str) -> bool:
    k = key.lower()
    if k in _TRACKING_PARAMS or k.startswith(_TRACKING_PREFIXES):
        return True
    return k in _HOST_TRACKING.get(host, ())


def url_key(url) -> str:
    """The canonical dedupe key for ``url`` ('' for anything that isn't an
    http(s) URL with a host). Pure, never raises."""
    if not isinstance(url, str):
        return ""
    raw = url.strip()
    if not raw:
        return ""
    try:
        parts = urlsplit(raw)
        if parts.scheme.lower() not in ("http", "https"):
            return ""
        host = _clean_host(parts.hostname or "")
        if not host:
            return ""
        port = parts.port
    except ValueError:
        return ""
    if port and port not in (80, 443):
        host = f"{host}:{port}"

    path = parts.path or ""
    params = parse_qsl(parts.query, keep_blank_values=True)

    yt = _youtube_id(host, path, params)
    if yt:
        return f"https://youtube.com/watch?v={yt}"

    # Re-quote so a percent-encoded and a raw path agree, and keep '/' as is.
    path = quote(path, safe="/:@!$&'()*+,;=-._~%")
    path = path.rstrip("/")

    kept = sorted((k, v) for k, v in params if not _is_tracking(k, host))
    query = urlencode(kept)
    return f"https://{host}{path}" + (f"?{query}" if query else "")

File: web/lib/urlKey.ts
/**
 * Canonical dedupe key for a saved URL (`urlKey` on a link doc).
 *
 * MIRRORS `functions/url_key.py` (`url_key`) rule for rule: the web form dedupes
 * against the same Firestore field the backend writes, so the two must produce
 * identical keys. functions/tests/test_url_key.py pins the cases; change both
 * files together.
 *
 * https always, host lower-cased with www./m./mobile. stripped and
 * twitter.com → x.com, YouTube short forms → youtube.com/watch?v=ID, no
 * fragment (except a hash route like `#/inbox` or `#!/post`, which IS the page
 * in a single-page app), no trailing slash, tracking params dropped, remaining params sorted.
 * Returns '' for anything that isn't an http(s) URL with a host.
 */

const TRACKING_PARAMS = new Set([
    'fbclid', 'gclid', 'dclid', 'gbraid', 'wbraid', 'msclkid', 'yclid',
    'twclid', 'ttclid', 'li_fat_id', 'igsh', 'igshid', 'ref', 'ref_src',
    'ref_url', '_ga', '_gl', 'mkt_tok', 'oly_anon_id', 'oly_enc_id',
    'vero_id', 'wickedid', 'rb_clickid', 's_cid', '__s', '_hsenc', '_hsmi',
    'spm', 'share_id', 'sharesource', 'cmpid',
]);
const TRACKING_PREFIXES = ['utm_', 'mc_', 'pk_', 'hsa_'];

const HOST_TRACKING: Record<string, Set<string>> = {
    'x.com': new Set(['s', 't']),
    'youtube.com': new Set(['si', 'feature', 'pp']),
    'open.spotify.com': new Set(['si', 'context', 'nd']),
    'spotify.com': new Set(['si']),
    'instagram.com': new Set(['img_index']),
};

const HOST_PREFIXES = ['www.', 'm.', 'mobile.'];
const HOST_ALIASES: Record<string, string> = { 'twitter.com': 'x.com' };
const YOUTUBE_HOSTS = ['youtube.com', 'youtube-nocookie.com'];
const YOUTUBE_PATH_IDS = ['/shorts/', '/live/', '/embed/', '/v/'];

function cleanHost(raw: string): string {
    let host = raw.trim().toLowerCase().replace(/\.+$/, '');
    for (const prefix of HOST_PREFIXES) {
        if (host.startsWith(prefix) && host.split('.').length - 1 >= 2) {
            host = host.slice(prefix.length);
            break;
        }
    }
    return HOST_ALIASES[host] ?? host;
}

function validYtId(value: string): string | null {
    const v = (value || '').trim();
    return v.length >= 6 && v.length <= 20 && /^[A-Za-z0-9_-]+$/.test(v) ? v : null;
}

function youtubeId(host: string, path: string, params: [string, string][]): string | null {
    if (host === 'youtu.be') return validYtId(path.replace(/^\/+|\/+$/g, '').split('/')[0]);
    if (YOUTUBE_HOSTS.includes(host)) {
        if (path.replace(/\/+$/, '') === '/watch') {
            const v = params.find(([k]) => k === 'v');
            if (v) return validYtId(v[1]);
        }
        for (const prefix of YOUTUBE_PATH_IDS) {
            if (path.startsWith(prefix)) return validYtId(path.slice(prefix.length).split('/')[0]);
        }
    }
    return null;
}

function isTracking(key: string, host: string): boolean {
    const k = key.toLowerCase();
    if (TRACKING_PARAMS.has(k) || TRACKING_PREFIXES.some(p => k.startsWith(p))) return true;
    return HOST_TRACKING[host]?.has(k) ?? false;
}

// Python's urllib.parse.quote with the safe set url_key.py uses.
function quotePath(path: string): string {
    return path.replace(/[^A-Za-z0-9/:@!$&'()*+,;=\-._~%]/g, c => {
        try { return encodeURIComponent(c); } catch { return c; }
    });
}

// A hash route (`#/…` or `#!…`) is kept, trailing slashes trimmed; any other
// fragment (an in-page anchor) is dropped. Mirrors url_key._route_fragment.
function routeFragment(hash: string): string {
    const frag = hash.startsWith('#') ? hash.slice(1) : hash;
    if (!frag || (frag[0] !== '/' && frag[0] !== '!')) return '';
    const trimmed = frag.replace(/\/+$/, '');
    if (trimmed === '' || trimmed === '!') return '';
    return trimmed.replace(/[^A-Za-z0-9/:@!$&'()*+,;=\-._~%?#]/g, c => {
        try { return encodeURIComponent(c); } catch { return c; }
    });
}

// Python's urlencode (quote_plus): space → '+', and !'()* escaped.
function quotePlus(s: string): string {
    return encodeURIComponent(s)
        .replace(/[!'()*]/g, c => '%' + c.charCodeAt(0).toString(16).toUpperCase())
        .replace(/%20/g, '+');
}

export function urlKey(input: unknown): string {
    if (typeof input !== 'string') return '';
    const raw = input.trim();
    if (!raw) return '';
    let parsed: URL;
    try {
        parsed = new URL(raw);
    } catch {
        return '';
    }
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') return '';
    let host = cleanHost(parsed.hostname);
    if (!host) return '';
    if (parsed.port && parsed.port !== '80' && parsed.port !== '443') host = `${host}:${parsed.port}`;

    const path = parsed.pathname;
    const params: [string, string][] = [];
    parsed.searchParams.forEach((v, k) => { params.push([k, v]); });

    const yt = youtubeId(host, path, params);
    if (yt) return `https://youtube.com/watch?v=${yt}`;

    const cleanPath = quotePath(path).replace(/\/+$/, '');
    const kept = params
        .filter(([k]) => !isTracking(k, host))
        .sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0));
    const query = kept.map(([k, v]) => `${quotePlus(k)}=${quotePlus(v)}`).join('&');
    const frag = routeFragment(parsed.hash);
    return `https://${host}${cleanPath}${query ? `?${query}` : ''}${frag ? `#${frag}` : ''}`;
}
